### DISCUSSION TITLE
Is there vfile-reporter-github?

### DISCUSSION THREAD

**Discussion by noomorph** [vfile/vfile#87] [Ideas]
GitHub Actions support this format:

https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-error-message

Is there any solution that fills that niche, or might you be interested in such a reporter (if someone contributes it)?

**Comment by wooorm** [comment]
Hi!

I don’t think there is yet, but it seems useful yeah!

Should it be an own package? Or should it be in https://github.com/vfile/vfile-reporter, and detect if GITHUB_CI (?) is in env, and switch to the GH specific output then?

**Reply by noomorph** [reply]
Well, I thought about a separate repo, but I don't know how it would work best for your ecosystem – wouldn't that bloat your `vfile-reporter`, I wonder? Generally, I'm okay with both options, but you are the boss here :) 

My use case is to set up a Github Action to validate documentation files in Markdown and MDX based on custom plugins (but you can think about spelling, etc.). That could be nice to see the spelling errors this way:

![](https://github.com/cschleiden/jest-github-actions-reporter/blob/master/img/annotations.png?raw=true)

P. S. 🤔 By the way, I just found someone's solution for that: https://github.com/cl8n/vfile-reporter-github-action
It won't fit your ecosystem as you rely on JSDoc, whereas the community plugin is TypeScript-based, but I think it can be reworked if needed.

**Reply by noomorph** [reply]
Alternatively, it could be a `vfile-reporter-ci`, perhaps? I mean if someone extends it with TeamCity Service Messages, and then adds some more CI providers, it would make still sense. Besides, every provider could be `export`ed separately.

**Reply by remcohaszing** [reply]
I definitely think this is great as a standalone package.

To include it in `vfile-reporter`, maybe? It feels a bit out of place, but it also seems useful. Should we also support other CI providers by default then? To me GitLab immediately comes to mind.

**Reply by noomorph** [reply]
@remcohaszing, I wonder, does the standard `vfile-reporter` provide a source code view to see what was wrong, just like `vfile-reporter-position`, e.g.:

![](https://raw.githubusercontent.com/Hocdoc/vfile-reporter-position/HEAD/screenshot.png)

I mean, I don't see such an option here:

![](https://github.com/vfile/vfile-reporter/blob/main/screenshot.png)

**Reply by noomorph** [reply]
What I imply is, as a user of `vfile` reporters, I still might not be satisfied with the default one, whereas that puts me into the position where I'd have to use the default one along with an alternative one, plus swap the selected reporter conditionally: use `vfile-reporter` (CI-only) and use another reporter (local run).

**Reply by remcohaszing** [reply]
I have been thinking it could be nice if the reporter can be specified using an environment variable. This way you could add this to your `.bashrc` / `.zshrc`:

```sh
export UNIFIED_ENGINE_REPORTER=vfile-reporter-position
```

And install `vfile-reporter-position` globally:

```sh
npm install -g vfile-reporter-position
```

---

Also have you seen [`remark-language-server`](https://github.com/remarkjs/remark-language-server) / [`remark for VSCode`](https://marketplace.visualstudio.com/items?itemName=unifiedjs.vscode-remark)?

**Reply by noomorph** [reply]
@remcohaszing, hm... I'm just looking at the public interface of `vfile-reporter`, and...

```js
import {VFile} from 'vfile'
import {reporter} from 'vfile-reporter'

const one = new VFile({path: 'test/fixture/1.js'})
const two = new VFile({path: 'test/fixture/2.js'})

one.message('Warning!', {line: 2, column: 4})

console.error(reporter([one, two]))
```

and, if it were an uncannily smart `process.env`-based export, I'd say that it is a code smell, no?

In that case, some extra library like `vfile-reporter-provider` would make more sense, IMO.

## So, getting back to the idea...

So, getting back to the idea with extending the default `vfile-reporter`, don't you think [the code there](https://github.com/vfile/vfile-reporter/blob/main/lib/index.js) is too "straightforward"? I mean, it does not imply any composition/decoration, etc. We'd have to refactor it quite aggressively to add such feature as CI reporting. 🤔 Do you really think this is the way to go? As a side contributor, I can't decide things like that, so it's up to the maintainers.

P. S. No, I'm not a VSCode guy, so I totally missed [remark-language-server](https://github.com/remarkjs/remark-language-server) / [remark for VSCode](https://marketplace.visualstudio.com/items?itemName=unifiedjs.vscode-remark). Indeed, that looks like another good solution. Thanks for pointing that out.

**Reply by wooorm** [reply]
So looks like there's a good reporter that does this already.
People who directly use this package, already specify which one they load, so they could just as well use that one, and with 1 line of code switch when running in GH CI.
On the unified-engine side, that sounds quite useful to automatically use GH when it detects that. And could use GL or whatever else too. But not sure if that's your question?

**Reply by noomorph** [reply]
@wooorm, it seems you slightly misunderstood the train of thought in this discussion (which is relatable since it is long and has been branching in several directions since the original question).

Long story short, I started a (currently private) repo, `@noomorph/vfile-reporter-ci` on basis of your `vfile/*` repos (same scripts, JSDoc, etc), but I'm not sure yet if it fits your standards. Do you think you could adopt a small repo similar to `vfile-reporter` (but for CI)?

If yes, I'll need your review on the test strategy for `vfile-reporter-ci`. I see you tend to use built-in `node:test` stuff, and you avoid testing frameworks, which slightly complicates my job here. An ideal test for a CI reporter is a Selenium WebDriver test which goes over the build page and checks that the CI correctly understands the output of that reporter. My gut feeling is that you say it is overkill for you to support, and I would even agree. Maybe, some basic snapshot tests would do the justice?

If not, and there's no chance you would adopt the repo, then perhaps I should not care about your opinion very much, but still, I think it would be nice to get my repo included under the official `vfile` org, so I'd try my chances. 😇 

P.S. I'm also unsure about how many CI providers are enough. I'm fine with 1 😄 , but I made a modular system, so volunteers can add more CI providers in separate PRs. Does it work for you? If not, the best extra mile I can go is, perhaps, TeamCity, but not beyond this.

**Reply by wooorm** [reply]
> Do you think you could adopt a small repo similar to vfile-reporter (but for CI)?

Yep!

> My gut feeling is that you say it is overkill for you to support, and I would even agree. Maybe, some basic snapshot tests would do the justice?

Yep too! Feels like a lot of complexity. Basic snapshots are fine with me

> Does it work for you?

Yep, I mostly care about GH. And if someone wants to add Gitlab or something else later that’s fine!

> but still, I think it would be nice to get my repo included under the official vfile org, so I'd try my chances.

Plugins, utilities, extensions, can live fine outside of our organizations!
When inside, it means that they are administered/maintained/published by our teams (https://github.com/unifiedjs/collective#vfile-team).
Practically, it likely means that I’ll refactor stuff to match the rest of the ecosystem, and go through things every other year or so.
You could also become a contributor/maintainer of vfile if you want!

**Reply by noomorph** [reply]
Okay, hopefully tomorrow I can share the repo.

**Reply by remcohaszing** [reply]
Another approach is to use [GitHub action problem matchers](https://github.com/actions/toolkit/blob/main/docs/problem-matchers.md). For example, `actions/setup-node` has [problem matchers for ESLint](https://github.com/actions/setup-node/blob/main/.github/eslint-stylish.json). Perhaps they’ll also accept one for the default vfile reporter too.

---

GitLab uses a different approach though. It requires writing to a file, not stdout.

**Reply by noomorph** [reply]
@remcohaszing this is exactly why it makes more sense to create specific reporters.

**Reply by noomorph** [reply]
See: https://github.com/noomorph/vfile-reporter-ci

**Reply by wooorm** [reply]
Looks like its working? Anything in particular you want me/us to look at?

**Reply by noomorph** [reply]
@wooorm, correct. I have a few things to discuss. See this commit and my comments inside it:
https://github.com/noomorph/vfile-reporter-ci/commit/32b2c2545d88bd1344d276883f88d0fe875a4cad

**Reply by wooorm** [reply]
left some comments!
